### ISSUE TITLE
company-cycle-threshold - direct insertion of candidates

### ISSUE THREAD

**Issue by minad** [company-mode/company-mode#1061]
I am using company with this configuration, mostly with capf:

~~~ elisp
  :bind (:map company-mode-map
         ("TAB" . company-indent-or-complete-common)
         :map company-active-map
         ("<tab>" . company-complete-common-or-cycle)
         ("TAB" . company-complete-common-or-cycle))
~~~

Is there a possibility to configure a `company-cycle-threshold` similar to the `completion-cycle-threshold`, which would directly insert the candidate if there exist no more candidates than the given threshold? If the threshold is exceeded, open the popup.

**Comment by dgutov** [comment]
I don't get it. How will it work?

Like, if there are no more than 5 completions, insert the first one?

**Comment by minad** [comment]
Yes. Completion at point works in the same way. TAB, Insert first candidate, TAB, replace with second, ... Only if more candidates exist, the completions buffer opens.

**Comment by dgutov** [comment]
You are using `company-tng-mode`, right?

You should be able to get this behavior by additionally replacing `company-pseudo-tooltip-unless-just-one-frontend` in `company-frontends` with a custom wrapper for `company-pseudo-tooltip-frontend`. One that will count the length of `company-candidates` first.

**Comment by minad** [comment]
Oh okay, I am not using tng mode, I missed totally that this exists. I have mainly this seemless shell tabbing in mind `cd /a TAB b TAB TAB c ...`  such that I don't have to press `RET` in between and such that the company menu does not come up unnecessarily for very few candidates. I will try tng mode and report back here! Thank you!

**Comment by minad** [comment]
Your suggestion seems to work well and is going in the right direction.

~~~ elisp
(defun company-pseudo-tooltip-threshold-frontend (command)
  (unless (and (eq command 'post-command)
               (<= (length company-candidates) completion-cycle-threshold))
    (company-pseudo-tooltip-frontend command)))

(setq company-frontends '(company-tng-frontend company-pseudo-tooltip-threshold-frontend company-echo-metadata-frontend))
~~~

I have a few remaining issues:

1. Since I have bound TAB to `company-indent-or-complete-common` I sometimes have to press TAB twice in order to see the candidate, in particular if there is only a single candidate. Then the first keypress completes the prefix (basically does nothing) and then the second keypress calls company-select-next since then the company-active-map sets in.

~~~ elisp
  :bind (:map company-mode-map
         ("TAB" . company-indent-or-complete-common)
         :map company-active-map
         ("<tab>" . company-select-next)
         ("<backtab>" . company-select-previous))
~~~

Is there some "more dwimish" variant of company-indent-or-complete-common, which completes if there is something to complete and if there is nothing to complete and the number of candidates is below the threshold, directy calls company-select-next?

2. TAB should always complete and only select the next candidate when in the low-threshold cycling mode. This means I also need some replacement for company-select-next and company-select-previous.

3. When starting with many candidates, pressing TAB shows the popup, then entering more input untilI am left with only three candidates. Then the popup is gone. Ideally the popup should never disappear until the final candidate has been selected. Basically the company-pseudo-tooltip-threshold-frontend criterion needs some additional override.

So to summarize - below the threshold company is in the cycling mode, TAB cycles. Above the threshold company is in the completion mode, TAB shows the popup and completes. The popup remains visible. And company never appears by itself, only by pressing TAB. This is what I would like to achieve :)

**Comment by minad** [comment]
This config implements what I had in mind:

~~~ elisp
(use-package company
  :hook ((prog-mode . company-mode)
         (eshell-mode . company-mode))
  :bind (:map company-mode-map
         ("TAB" . my--company-indent-or-complete-common)
         :map company-active-map
         ("<tab>" . my--company-complete-or-next))
  :config
  (company-tng-mode)

  (defvar-local my--company-popup nil)

  (defun my--company-indent-or-complete-common ()
    (interactive)
    (setq-local my--company-popup nil)
    (call-interactively #'company-indent-or-complete-common)
    (unless (and company-common (cdr company-candidates))
      (my--company-complete-or-next)))

  (defun my--company-pseudo-tooltip-threshold-frontend (command)
    (let ((low (<= (length company-candidates) completion-cycle-threshold)))
      (unless (and (eq command 'post-command) (not my--company-popup) low)
        (company-pseudo-tooltip-frontend command)
        (unless low (setq-local my--company-popup t)))))

  (defun my--company-complete-or-next ()
    (interactive)
    (call-interactively
     (if (<= (length company-candidates) completion-cycle-threshold)
         #'company-select-next
       #'company-complete-selection)))

  (setq company-frontends '(company-tng-frontend
                            my--company-pseudo-tooltip-threshold-frontend
                            company-echo-metadata-frontend))

  (setq company-idle-delay nil
        company-selection-wrap-around t
        company-backends '(company-capf)))
~~~

**Comment by minad** [comment]
@dgutov I am closing this since for me the issue has been resolved!

**Comment by dgutov** [comment]
Very good.

I will note, though, that having the "complete-common" behavior together with tng is pretty unusual. `company-tng-mode` implements the behavior from the "other" school of thought (ycmd and friends) where you just type, have completions fuzzy matched, and press TAB if you see the completion you like.

But of course it's good to be able to mix behaviors if one prefers.

**Comment by minad** [comment]
Yes, I know it is weird. I am trying to reconcile somehow the tabbing completion I am used to from shells with the fuzzy completion of the minibuffer. For reference I can paste my full config later. Maybe it turns out that I won't use this for long. It is an experiment.

**Comment by dgutov** [comment]
Oh, BTW you can also try `company-pseudo-tooltip-unless-just-one-frontend-with-delay`.

It introduces a delay before the popup is shown instead of basing the logic on the total number of completions.

**Comment by minad** [comment]
Yes, the delay would work out too. I've seen that frontend. But right now the configuration is to use tabbing for `candidates <= threshold` and for more candidates to use minibuffer like completion with the popup. By not showing the popup at all for few candidates the behavior is somehow visually separated.

**Comment by minad** [comment]
This is the current status:

~~~ elisp
(use-package company
  :hook ((prog-mode . company-mode)
         (eshell-mode . company-mode))
  ;; Use company for TAB instead of completion-at-point
  :bind (:map company-mode-map
         ("TAB" . dm/company-indent-or-complete-common)
         :map company-active-map
         ("<tab>" . dm/company-complete-or-next)
         ("<return>" . dm/company-finish)
         ("SPC" . dm/company-space))
  :config

  (require 'company-tng)

  (defvar dm/company-popup-opened nil)

  (defun dm/company-finish ()
    "Finish completion. Append additional RET if popup is open."
    (interactive)
    (setq company-selection (or company-selection 0))
    (company-complete-selection)
    (unless dm/company-popup-opened
      (setq unread-command-events (append unread-command-events '(13)))))

  (defun dm/company-space ()
    "Insert some unicode letter character, see also orderless-component-separator."
    (interactive)
    (insert (propertize "𖫩" 'display " ")))

  ;; See https://github.com/company-mode/company-mode/issues/1061
  (defun dm/company-indent-or-complete-common ()
    "Indent or complete. When no common prefix has been completed, insert candidate."
    (interactive)
    (setq-local dm/company-popup-opened nil)
    (let ((pos (point))
          (ind (current-indentation)))
      (call-interactively #'company-indent-or-complete-common)
      ;; Insert candidate if no prefix completion happened
      ;; dm/company-complete-or-next only selects the candidate
      ;; if fewer candidates than threshold.
      (when (and (= pos (point)) (= ind (current-indentation)))
        (dm/company-complete-or-next))))

  (defun dm/company-pseudo-tooltip-threshold-frontend (command)
    "Only show company popup if there are fewer than completion-cycle-threshold candidates."
    (let ((low (<= (length company-candidates) completion-cycle-threshold)))
      (unless (and (eq command 'post-command) (not dm/company-popup-opened) low)
        (company-pseudo-tooltip-frontend command)
        (unless low (setq-local dm/company-popup-opened t)))))

  (defun dm/company-complete-or-next ()
    "Select next candidate if there are fewer than completion-cycle-threshold candidates.
Otherwise complete prefix."
    (interactive)
    (call-interactively
     (if (<= (length company-candidates) completion-cycle-threshold)
         #'company-select-next
       #'company-complete-selection)))

  (defvar dm/company-changes nil)
  (defun dm/company-hook-start (manual)
    (when manual
      (activate-change-group (setq dm/company-changes (prepare-change-group)))))
  (defun dm/company-hook-cancel (manual)
    (when (and manual dm/company-changes)
      (cancel-change-group dm/company-changes)))
  (defun dm/company-hook-finish (manual)
    (when (and manual dm/company-changes)
      (accept-change-group dm/company-changes)))
  (add-hook 'company-completion-started-hook #'dm/company-hook-start)
  (add-hook 'company-completion-cancelled-hook #'dm/company-hook-cancel)
  (add-hook 'company-completion-finished-hook #'dm/company-hook-finish)

  ;; See https://github.com/oantolin/orderless#company
  (defun dm/company-adjust-orderless-faces (fn &rest args)
    "Fix company highlighting for orderless."
    (let ((orderless-match-faces [completions-common-part]))
      (apply fn args)))
  (advice-add 'company-capf--candidates :around #'dm/company-adjust-orderless-faces)

  (setq-default company-frontends '(company-tng-frontend
                                    dm/company-pseudo-tooltip-threshold-frontend
                                    ;;company-echo-metadata-frontend
                                    )
                company-selection-default nil ;; needed such that TAB/RET does not insert directly
                company-selection-wrap-around t ;; cycling
                company-idle-delay nil ;; no automatic completion
                company-backends '(;; company-nxml - use completion-at-point instead
                                   ;; company-css - use completion-at-point instead
                                   ;; company-semantic - not using semantic/cedet for now
                                   ;; company-cmake - not using cmake for now
                                   company-capf
                                   ;; company-clang - prefer lsp over the builtin clang completion
                                   company-files
                                   (company-dabbrev-code company-gtags company-keywords) ;; disabled company-etags
                                   company-dabbrev
                                   ;; company-elisp - use completion-at-point instead
                                   ;; company-bbdb
                                   ;; company-oddmuse
                                   )))
~~~

It looks all pretty intricate. Company seems to be arbitrarily hackable. If you have ideas for simplification, I would appreciate it. With such configurations I am always worried that they will break sooner or later.

**Comment by dgutov** [comment]
Do you really need `dm/company-finish`? Setting `company-selection-default` should already give a similar benefit, while making the popup less obtrusive (it won't insert completions unless you interact with it first).

Regarding it being intricate, well, you have some peculiar requirements around the "threshold" concept, so it should make sense that you'd have to make some effort to satisfy them (I don't expect this to be a very popular requirement). But suggestions how to make things simpler/reliable are welcome.

**Comment by minad** [comment]
I think I added this finish function for shell completion such that I don't have to press RET twice. But I have to check again what I wrote here. I am using this for a few days now and it seems to work well. If I come up with improvement proposals I let you know.
